src/vm/checker/typecheck/mod.rs

```diff
@@ -0,0 +1,404 @@
+        for arg in args.iter() {
+            // don't use map here, since type_check has side-effects.
+            result.push(self.type_check(arg, context)?)
+        }
+        Ok(result)
+    }
+
+    fn type_check_function_type(&mut self, func_type: &FunctionType,
+                                args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+        let typed_args = self.type_check_all(args, context)?;
+        func_type.check_args(&typed_args)?;
+        Ok(func_type.return_type().clone())
+    }
+
+    fn type_check_list_pairs<'b> (&mut self, bindings: &[SymbolicExpression],
+                                  context: &'b TypingContext) -> CheckResult<TypingContext<'b>> {
+        let mut out_context = context.extend()?;
+        for binding in bindings.iter() {
+            let binding_exps = binding.match_list()
+                .ok_or(CheckError::new(CheckErrors::BadSyntaxBinding))?;
+            
+            if binding_exps.len() != 2 {
+                return Err(CheckError::new(CheckErrors::BadSyntaxBinding))
+            }
+
+            let var_name = binding_exps[0].match_atom()
+                .ok_or(CheckError::new(CheckErrors::BadSyntaxBinding))?;
+
+            self.type_map.set_type(&binding_exps[0], no_type())?;
+            let typed_result = self.type_check(&binding_exps[1], context)?;
+            out_context.variable_types.insert(var_name.clone(),
+                                              typed_result);
+        }
+
+        Ok(out_context)
+    }
+
+    fn get_function_type(&self, function_name: &str) -> Option<FunctionType> {
+        if let Some(function_type) = self.contract_context.get_function_type(function_name) {
+            Some(function_type.clone())
+        } else {
+            None
+        }
+    }
+
+    fn type_check_define_function(&mut self, function_expression: &[SymbolicExpression],
+                                  context: &TypingContext) -> CheckResult<(String, FunctionType)> {
+        if function_expression.len() != 3 {
+            return Err(CheckError::new(CheckErrors::IncorrectArgumentCount(2, function_expression.len() - 1)))
+        }
+
+        self.type_map.set_type(&function_expression[0], no_type())?;
+        self.type_map.set_type(&function_expression[1], no_type())?;
+        // should we set the type of the subexpressions of the signature to no-type as well?
+
+        let signature = function_expression[1].match_list()
+            .ok_or(CheckError::new(CheckErrors::DefineFunctionBadSignature))?;
+        let body = &function_expression[2];
+
+        let (function_name, args) = signature.split_first()
+            .ok_or(CheckError::new(CheckErrors::VariadicNeedsOneArgument))?;
+        let function_name = function_name.match_atom()
+            .ok_or(CheckError::new(CheckErrors::BadFunctionName))?;
+        let mut args = parse_name_type_pairs(args)
+            .map_err(|_| { CheckError::new(CheckErrors::BadSyntaxBinding) })?;
```

Comment:
When I was trying out the VM, I didn't know what this error meant.  Could this error arise from a simple typo, or (in my case) a misapplication of Lisp syntax?

Reply:
Yeah -- this error message could definitely be more helpful. It could happen due to simple typos or misapplication of syntax.